### DISCUSSION TITLE
simple lua combobox sysex

### DISCUSSION THREAD

**Discussion by howidoo** [RomanKubiak/ctrlr#341] [General]
hi everyone
I want to make a simple combox for example:

Reverb 
Hall1
Hall2

every setting has a different sysex string i want to send via lua. 
May someone can show me how to write this. That would be great.
I found this code but it doesnt work 

`
-- Called when a modulator value changes
-- @mod   http://ctrlr.org/api/class_ctrlr_modulator.html
-- @value    new numeric value of the modulator
--
rev = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)

if
combobox == 1 then
SendMIDINow(F0 43 10 4C 02 01 00 10 00 F7)
elseif
combobox == 2 then
SendMIDINow(F0 43 10 4C 05 01 00 10 01 F7)
end`

thanks for your help

**Comment by dobo365** [comment]
Hi, please try this one instead ;-)

--
-- Called when a modulator value changes
-- @mod   http://ctrlr.org/api/class_ctrlr_modulator.html
-- @value    new numeric value of the modulator
--
rev = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)

	-- No action if the panel is in bootstrap or program states
	if panel:getBootstrapState() == true or source==6 or source==1 then
		return
	end	

	msg = CtrlrMidiMessage({0xF0, 0x43, 0x10, 0x4C, 0x02, 0x01, 0x00, 0x10, value, 0xF7})
	panel:sendMidiMessageNow(msg)

end

**Reply by howidoo** [reply]
hey thanks for your replay. The code worked :)
But I need to change more then one value, Is it possible to change msb and lsb with the combobox


F0 43 10 4c 02 01 00 02 00 F7
F0 43 10 4c 02 01 00 03 01 F7

thank you

**Reply by damiensellier** [reply]
Everything is possible : )
... 02 00 ...
... 03 01 ...
... 04 02 ...
... 05 03 ...

Is the scheme always the same like ...MSB+1, LSB+1... ? If so 

```
msg = CtrlrMidiMessage({0xF0, 0x43, 0x10, 0x4c, 0x02, 0x01, 0x00, value+2, value, 0xF7})
panel:sendMidiMessageNow(msg)
```

**Comment by howidoo** [comment]
![lsb msb](https://user-images.githubusercontent.com/88279503/127785025-a3bf3eee-0220-42c0-b6d1-4a94d1efdfa7.jpg)

**Reply by howidoo** [reply]
It works when i use buttons and send the full sysex string. I also tried
F0 43 10 4c 02 01 00 LS MS F7 without Lua
and write for instance: Hall2=257 in the combobox. But when i look at the Midimonitor it sends RAW:[f0 43 10 4c 02 01 00 01 02 f7] not RAW:[f0 43 10 4c 02 01 00 01 01 f7]

**Comment by dnaldoog** [comment]
Here is a panel that doesn't use _lua_.

The _sysex_ formula is `F0 43 10 4C 02 01 00 MS LS F7` and the combo contents are:

<pre>
NO EFFECT=0
HALL1=128
HALL2=129
ROOM1=256
ROOM2=257
ROOM3=258
STAGE1=384
STAGE2=385
PLATE=512
WHITE ROOM=1280
TUNNEL=1408
BASEMENT=1664
</pre>

:)
----
[COMBO REVERB NO lua_1.0.bpanelz.zip](https://github.com/RomanKubiak/ctrlr/files/6913627/COMBO.REVERB.NO.lua_1.0.bpanelz.zip)

**Comment by dnaldoog** [comment]
Here is a panel that uses _lua_.

The code is:

<pre>
send_reverb = function(--[[ CtrlrModulator --]] mod --[[ number --]], value --[[ number --]], source)

    local bi = BigInteger(mod:getValueMapped())
    local msb = bi:getBitRangeAsInt(7, 14)
    local lsb = bi:getBitRangeAsInt(0, 7)
    local sysex = string.format("F0 43 10 4C 02 01 00 %.2x %.2x F7", msb, lsb)
    local m = CtrlrMidiMessage(sysex)
    panel:sendMidiMessageNow(m)

end

</pre>
----

You could also use a lua lookup table in the function mapped to  `value`.

<pre>
send_reverb = function(--[[ CtrlrModulator --]] mod --[[ number --]], value --[[ number --]], source)
    local t={[0]=0 ,128 ,129 ,256 ,257 ,258 ,384 ,385 ,512 ,1280 ,1408 ,1664}
    local bi = BigInteger(t[value])
    local msb = bi:getBitRangeAsInt(7, 14)
    local lsb = bi:getBitRangeAsInt(0, 7)
    local sysex = string.format("F0 43 10 4C 02 01 00 %.2x %.2x F7", msb, lsb)
    local m = CtrlrMidiMessage(sysex)
    panel:sendMidiMessageNow(m)

end

</pre>
----

[COMBO REVERB with lua_1.0.bpanelz.zip](https://github.com/RomanKubiak/ctrlr/files/6913756/COMBO.REVERB.with.lua_1.0.bpanelz.zip)

**Reply by howidoo** [reply]
great thanks finally it worked:)

**Reply by damiensellier** [reply]
@howidoo can you please tick the thread as solved?
